File: spec_to_review/lib/courier/bell/notification.rb


=begin

Copyright (c) 2021, all rights reserved.

All the information provided by this platform is protected by international laws related  to
industrial property, intellectual property, copyright and relative international laws.
All intellectual or industrial property rights of the code, texts, trade mark, design,
pictures and any other information belongs to the owner of this platform.

Without the written permission of the owner, any replication, modification,
transmission, publication is strictly forbidden.

For more information read the license file including with this software.

// · ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~
// ·

=end

require "rails_helper"
require "spec_helper"
require "byebug"

RSpec.describe Courier::Bell::Notification, type: :model do

    before(:all) do
        return unless defined? CloudBell

        @account = User.find_by(email: Rails.application.config.lesli_settings["account"]["user"]["email"]).account # instance of account

        # create user
        password = Faker::Internet.password
        @user = @account.users.create({
            email: Faker::Internet.email,
            password: password,
            password_confirmation: password,
            detail_attributes: {
                first_name: Faker::Name.first_name,
                last_name: Faker::Name.last_name,
                salutation: ["mr", "mrs"][rand(2)],
                address: Faker::Address.city
            }
        })
    end

    it "Courier::Bell::Notification.count" do
        expect(Courier::Bell::Notification.count(@user)).to eq(@account.bell.notifications.count)
    end

    it "Courier::Bell::Notification.index" do
        notifications =  Courier::Bell::Notification.index(@user, query())

        # check pagination
        expect(notifications[:pagination][:count_total]).to eq(@account.bell.notifications.count)
        expect(notifications[:pagination][:current_page]).to eq(query()[:pagination][:page])
        expect(notifications[:pagination][:count_results]).to eq(query()[:pagination][:perPage])

        # check data
        expect(notifications[:records].class).to eq(Array)
    end

    it "Courier::Bell::Notification.read" do
        notification = @account.bell.notifications.create(
            user: @user,
            subject: Faker::Lorem.word,
            body: Faker::Lorem.sentence,
            url: Faker::Internet.url,
            category: "info",
            channel: "web",
            status: "created"
        )


        # call read method
        CloudBell::Notification.read(User.first,  notification.id)

        expect(CloudBell::Notification.find_by(id: notification.id).status).to eq(CloudBell::Notification.statuses[:read])
    end

    it "Courier::Bell::Notification.new" do
        notification = Courier::Bell::Notification.new(
            @user,
            Faker::Lorem.word,
            body: Faker::Lorem.sentence,
            url: Faker::Internet.url,
            category: "info",
            channel: "web"
        )

        expect(notification[:id]&.last).to be >= 0
    end

    private

    def query()
        @query = {
            filters: {},
            pagination: {
                perPage: 15,
                page: 1,
                order: "desc",
                orderColumn: "id"
            }
        }
    end

    def random_channel()
        channels = CloudBell::Notification.channels

        return channels[rand(channels.length)]
    end

    def random_category()
        categories = CloudBell::Notification.categories

        return categories[rand(categories.length)]
    end
end

File: lib/vue3/stores/shared/cloud_bell/notification.js
/*
Copyright (c) 2023, all rights reserved.

All the information provided by this platform is protected by international laws related  to
industrial property, intellectual property, copyright and relative international laws.
All intellectual or industrial property rights of the code, texts, trade mark, design,
pictures and any other information belongs to the owner of this platform.

Without the written permission of the owner, any replication, modification,
transmission, publication is strictly forbidden.

For more information read the license file including with this software.

// · ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~     ~·~
// ·
*/

// · This store is needed to make http requests, manage the reactivity
// · of a workflow.

import { defineStore } from "pinia"

export const useNotification = defineStore("core.shared.bell.notification", {
    state: () => {
        return {
            loading: false,
            notification: {
                subject: "",
                body: "",
                channel: "",
                user_receiver_id: ""
            },
            options: {
                categories: [],
                channels: [],
            }
        }
    },
    actions: {
        /**
         * @description This action is used to send a Notification
         */
        sendNotification(){
            this.loading = true

            this.http.post(this.url.bell("notifications"), {
                notification: this.notification
            }).then(() => {
                this.msg.success(I18n.t("core.users.messages_success_operation"))
            }).catch(error => {
                this.msg.danger(I18n.t("core.shared.messages_danger_internal_error"))
            }).finally(() => {
                this.loading = false
                this.notification = {
                    subject: "",
                    body: "",
                    channel: "",
                    user_receiver_id: ""
                }
            })

        },

        /**
         * @description This action is used to get options for notifications
         */
        getOptions(){
            this.options = {
                categories: [],
                channels: [],
            }

            this.loading = true

            this.http.get(this.url.bell("notifications/options")).then(result => {

                result["categories"].forEach((category)=>{
                    this.options["categories"].push({
                        label: category.text,
                        value: category.value
                    })
                })

                result["channels"].forEach((channel)=>{
                    this.options["channels"].push({
                        label: channel.text,
                        value: channel.value
                    })
                })

            }).catch(error => {
                this.msg.danger(I18n.t("core.shared.messages_danger_internal_error"))
            }).finally(()=>{
                this.loading = false
            })
        },


    }
})
